### DISCUSSION TITLE
Currying considered harmful + significant parens heresy feature

### DISCUSSION THREAD

**Discussion by Ericson2314** [ghc-proposals/ghc-proposals#487] [Ideas]
First of, I apologize for the sins I am about to commit :).

So, it is fairly known that currying is damn convenient, but also a frequent source of obscure errors (https://github.com/haskell/error-messages/issues/43 is but one example).

I think currying is most useful application side --- binding side is more rare and I think it is easier to without it there. This observation opens the door to a funny language extension. I view this as too ugly to imagine it getting in any future Haskell report, but it could be useful in a teaching language.

So without further ado:

## `-XNoPartiallyCurriedBindings`

The idea is that any binding form *with an explicit signature* should require all or no bindings up to the first paren:

```haskell
f :: Int -> Bool -> IO ()
f x b = ... -- legal
```
```haskell
f :: Int -> Bool -> IO ()
f x = ... -- illegal, didn't match the second arg
```
```haskell
f :: Int -> (Bool -> IO ())
f x = ... -- legal
f x = \b -> ... -- legal, mere special case of the above
```
```haskell
f :: Int -> (Bool -> IO ())
f x b = ... -- illegal, matched args of the "returned function"
```
```haskell
(\x b -> ...) :: Int -> Bool -> IO () -- legal
```
```haskell
(\x -> ...) :: Int -> Bool -> IO () -- illegal -- illegal, didn't match the second arg
```
```haskell
(\x -> ...) :: Int -> (Bool -> IO ()) -- legal
(\x -> \b -> ...) :: Int -> (Bool -> IO ()) -- legal, mere special case of the above
```
```haskell
(\x b -> ...) :: Int -> (Bool -> IO ()) -- illegal, matched args of the "returned function"
```

TODO extend to `@` binders of invisible stuff. `(..)` binder from #448 is always OK.

## Type checking ramifications

With a program like 
```haskell
foo :: Int -> Bool -> IO ()
foo i = do { ... ; putStrLn "hi!" }
```
we know we can exclude
```haskell
do { ... ; putStrLn "hi!" } :: Bool -> IO ()
```
as a possible solution
and so we type check as 
```haskell
foo :: Int -> Bool -> IO ()
foo i = (do { ... ; putStrLn "hi!" }) :: IO ()
```
This, without violating "guess freeness" makes GHC "assign blame" in a way more likely to reflect the user's intent. GHC will blame `foo` for not having the right number of arguments, rather than the `do` block for not being the `(->) Bool` Monad.

Just banning currying (lambdas and fun binds only take single argument, parens for each function application required, use a tuple if you want more stuff at a time) is far more principled, but makes it impossible to ergonomically use existing libraries.

This is a middle ground where existing libraries are totally useable, most idiomatic code passes, but the rules are yucky.

Still, maybe in a https://terrytao.wordpress.com/career-advice/theres-more-to-mathematics-than-rigour-and-proofs/ way this is good as an intermediate language for beginners (after no currying) and experts for which the ad-hoc-ness won't rot an already enlightened brain. It's just the "advanced beginners" who must experience what life is like without ad-hoc rules that are harmed by using this extension.

## This is crazy! Is there precedent?

I think https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/explicit_forall.html#the-forall-or-nothing-rule is arguably precedent. It can also be subverted with parens:

```haskell
const :: (forall a. b -> a -> b)
const a b = a
```

is in fact valid

In https://github.com/ghc-proposals/ghc-proposals/pull/285 I argued the forall or nothing rule is icky and I don't like extra complexity. I would say this feature is an "unsavory convenience" in a very similar manor to `-XImplicitForAll`.

**Comment by JakobBruenker** [comment]
What about
```haskell
f :: Int -> (Bool -> IO ())
f x b = ...
```
I imagine this would be illegal, too?

**Reply by Ericson2314** [reply]
Yeah, I guess so, good point.

**Reply by Ericson2314** [reply]
Updated first post

**Comment by noughtmare** [comment]
What about this:

```haskell
f :: Int -> Bool -> IO ()
f = \x y -> ... 
```

**Reply by Ericson2314** [reply]
I think the lambda is "far enough" away from the signature that the restriction shouldn't apply. That leaves `= \ .. -> ..` as an marked "escape hatch" too.

**Comment by nomeata** [comment]
What about `f = \case …` ;-)

**Reply by Ericson2314** [reply]
Ahahah.

In seriousness, that does help ensure one is writing a single-param lambda, but doesn't help with the problem that GHC always assumes the explicit arity is in the right and the body in the wrong when there are too few parameters.

E.g.:
```haskell
Prelude> (\x y -> ()) :: () -> () -> () -> ()
```
```
<interactive>:4:10: error:
    • Couldn't match expected type ‘() -> ()’ with actual type ‘()’
    • In the expression: ()
      In the expression: (\ x y -> ()) :: () -> () -> () -> ()
      In an equation for ‘it’: it = (\ x y -> ()) :: () -> () -> () -> ()
```

The ideal thing could be a super magic backtracking solver for HIE to figure out what the "best" way to assign blame is.

This is sort of a simpler trick, also aiming at helping people read code without "typechecker assistance", to blame the bindings arity instead of the body, in the one case where we're reasonably confident that's the right way to assign blame.

**Reply by Ericson2314** [reply]
See https://github.com/haskell/error-messages/issues/43 for a more real world example.

**Reply by Ericson2314** [reply]
I could also add, even if `\case` did help somehow, having people type more characters to get better errors is perverse incentivization, but that would be me not getting the joke :).

**Comment by konsumlamm** [comment]
One common use case that would be forbidden with this extension (afaict) is defining an alias, e.g. if you want an operator version and a "normal function" version (something like `cons = (:)`). I can't imagine such an extension being useful anyway outside of early teaching (sooner or later beginners have to learn about currying). In particular, I feel very uncomfortable about parentheses in type signatures changing the meaning of the types. This makes me wonder, how would this work with higher order functions? Would they need to be fully applied too? Or would they not be affected?

A better path to take might be writing a custom GHC plugin (disclaimer: I have no idea how these work) intended for teaching, that changes some stuff like this to help beginners. In my opinion, a GHC extension is the wrong way to go for this purpose. (I'm not sure how serious this proposal is anyway.)

That being said, I forgive you your sins :).

**Reply by Ericson2314** [reply]
`cons = (:)` is completely allowed, as is any "point free" stuff.

> In particular, I feel very uncomfortable about parentheses in type signatures changing the meaning of the types. 

Yes me too, but I don't think it is the meaning of the *type* in a very deep sense. It's easy to subvert this restriction by not putting the binder right up against the type:

```haskell
(\x -> ...) :: Int -> Bool -> ()
```
I think it's fine because the parens on the LHS prevent the restrictions from applying.

> This makes me wonder, how would this work with higher order functions? Would they need to be fully applied too? Or would they not be affected?

Not be effected. This is similar to the "forall or nothing rule" I just realized. Modified the top part to say so.

> A better path to take might be writing a custom GHC plugin (disclaimer: I have no idea how these work) intended for teaching, that changes some stuff like this to help beginners.

Hmm, yes I suppose that should be possible. 

> In my opinion, a GHC extension is the wrong way to go for this purpose. (I'm not sure how serious this proposal is anyway.)

I am also unsure! I am all for teaching languages, but in general I like doing elegant `-XNo` proposals that rip out features and make programs more verbose but more explicit. If plugins + HLS is not an ops nightmare, I suppose that can suffice for teaching.

**Comment by nomeata** [comment]
I am very hesitant. What I love about Haskell is the ability to abstract everywhere and easily. So going from

```
foo :: [T] -> [T]
foo = cons (bar 5)
```
to
```
foo' :: Int -> [T] -> [T]
foo' n = cons (bar n)
```
should be easily possible. Same for wrapping and unwrapping function types in `type` aliases.

Haskell without Curry is maybe a cat, but no longer the Haskell we know :-)

**Comment by Boarders** [comment]
Unless this were accompanied with a wider perspective around racket-style "language levels" (which I think it would be nice as a goal, though hard to conceptualize) or some greater pedagogical project, then I feel the medicine would be worse than the poison.
